service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,240 @@
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in getGiftCertificate(String name): " + e.getMessage());
+            throw new ServiceException("Failed to get certificate by it name: " + name,
+                    ErrorCodeEnum.FAILED_TO_RETRIEVE_CERTIFICATE);
+        }
+    }
+
+    @Override
+    public GiftCertificate getGiftCertificate(int id) throws ServiceException {
+        certificateValidator.validateId(id);
+        try {
+            return giftCertificateDAO.getGiftCertificate(id);
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in getGiftCertificate(int id): " + e.getMessage());
+            throw new ServiceException("Failed to get certificate by it id: " + id,
+                    ErrorCodeEnum.FAILED_TO_RETRIEVE_CERTIFICATE);
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> geAllCertificates() throws ServiceException {
+        try {
+            return giftCertificateDAO.getAllGiftCertificates();
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in getAllGiftCertificates(): " + e.getMessage());
+            throw new ServiceException("Failed to get all certificates", ErrorCodeEnum.FAILED_TO_RETRIEVE_CERTIFICATE);
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getGiftCertificatesByContent(String content) throws ServiceException {
+        if (content == null || content.isEmpty()) {
+            throw new ServiceException("Failed to validate: content is empty", ErrorCodeEnum.INVALID_INPUT);
+        }
+
+        try {
+            return giftCertificateDAO.getAllGiftCertificates(content);
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in getAllGiftCertificates(String content): " + e.getMessage());
+            throw new ServiceException("Failed to get certificate by it content: " + content,
+                    ErrorCodeEnum.FAILED_TO_RETRIEVE_CERTIFICATE);
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getGiftCertificateByTagName(String tagName) throws ServiceException {
+        certificateValidator.validateName(tagName);
+        try {
+            return giftCertificateDAO.getGiftCertificateByTagName(tagName);
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in getGiftCertificateByTagName(): " + e.getMessage());
+            throw new ServiceException("Failed to get certificate Failed to get certificate by tag name: " + tagName,
+                    ErrorCodeEnum.FAILED_TO_RETRIEVE_CERTIFICATE);
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getAllGiftCertificatesSortedByName(boolean isAscending)
+            throws ServiceException {
+        try {
+            return giftCertificateDAO.getAllGiftCertificatesSortedByName(isAscending);
```

Comment:
Sorting in service layer

Reply:
Same answer as for method getAllGiftCertificatesSortedByName.